docs/async-await-proposal.md

```diff
@@ -0,0 +1,442 @@
+`update`:
+
+```swift
+// Update is a bidirectional streaming call; provide a response handler.
+let update = client.update { response in
+  print("update received: \(response.text)")
+}
+
+// Send a bunch of messages to the service.
+for word in ["boyle", "jeffers", "holt"] {
+  let request = Echo_EchoRequest.with { $0.text = word }
+  update.sendMessage(request, promise: nil)
+}
+
+// Close the request stream.
+update.sendEnd(promise: nil)
+
+// wait() for the call to terminate
+let status = try update.status.wait()
+print("update completed with status: \(status.code)")
+```
+
+This API also exposes a number incidental types and patterns that the user need
+concern themselves with that are not specific to their application:
+
+1. It exposes the NIO types to the user, allowing the provision of an
+   `EventLoopPromise` when sending messages and requiring the use of
+   `EventLoopFuture` to obtain the `status` of the call.
+2. Code does not read in a straight line due to the need to provide a completion
+   handler when making the call.
+
+## Proposed solution
+
+### Proposed server API
+
+We propose generating the following new protocol which the user must conform to in
+order to implement the server:
+
+```swift
+/// To build a server, implement a class that conforms to this protocol.
+public protocol Echo_AsyncEchoProvider: CallHandlerProvider {
+  var interceptors: Echo_EchoServerInterceptorFactoryProtocol? { get }
+
+  /// Immediately returns an echo of a request.
+  func get(
+    request: Echo_EchoRequest,
+    context: AsyncServerCallContext
+  ) async throws -> Echo_EchoResponse
+
+  /// Splits a request into words and returns each word in a stream of messages.
+  func expand(
+    request: Echo_EchoRequest,
+    responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+    context: AsyncServerCallContext
+  ) async throws
+
+  /// Collects a stream of messages and returns them concatenated when the caller closes.
+  func collect(
+    requests: GRPCAsyncStream<Echo_EchoRequest>,
+    context: AsyncServerCallContext
+  ) async throws -> Echo_EchoResponse
+
+  /// Streams back messages as they are received in an input stream.
+  func update(
+    requests: GRPCAsyncStream<Echo_EchoRequest>,
+    responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+    context: AsyncServerCallContext
+  ) async throws
+}
+```
+
+Here is an example implementation of the bidirectional streaming `update`
+handler using this new API:
+
+```swift
+public func update(
+  requests: GRPCAsyncStream<Echo_EchoRequest>,
+  responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+  context: AsyncServerCallContext
+) async throws {
+  var count = 0
+  for try await request in requests {
+    let response = Echo_EchoResponse.with {
+      $0.text = "Swift echo update (\(count)): \(request.text)"
+    }
+    count += 1
+    try await responseStreamWriter.sendResponse(response)
+  }
+}
```

Comment:
Hm now that I think about it (after the comment above) I guess the writer is easier here... but you may want to figure out how you want to implement the backpressure here -- would you want to suspend here until more demand is abailable to push more responses? Internally I guess this would be implemented the ill named but final `AsyncStream` API, so you'll have all the tools to notice backpressure on that boundary, so that's good news 👍

Reply:
Stream writer probably handles backpressure by refusing to return from the write call when the window is full. This doesn't prevent multithreaded writing to the stream writer so we need to tolerate that, but this still works pretty well.

Also, the stream writer is a base abstraction: we can absolutely use it to implement an API that accepts an AsyncSequence.

Reply:
> Stream writer probably handles backpressure by refusing to return from the write call when the window is full.

Ok good 👍 

> Also, the stream writer is a base abstraction: we can absolutely use it to implement an API that accepts an AsyncSequence.

Yeah, good -- I saw that bit later on in the proposal, happy to see it 👍

Reply:
I could add a sentence or two about backpressure to this doc. As you noted, there is no attention given to it. This was a deliberate omission from the prototype because the underlying NIO gRPC library doesn't handle backpressure today. But, as discussed in this thread, we should be able to accommodate backpressure with this API.

Reply:
I added a note about (the lack of) backpressure support in https://github.com/grpc/grpc-swift/pull/1231/commits/71d3728062afe2afd850f92df825650e5cdb813a.